File: freightmas/trucking_service/report/truck_register/truck_register.py
# Copyright (c) 2025, Zvomaita Technologies (Pvt) Ltd and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {
            "label": _("Truck ID"),
            "fieldname": "name",
            "fieldtype": "Link",
            "options": "Truck",
            "width": 100,
        },
        {
            "label": _("Horse"),
            "fieldname": "horse",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": _("Trailer"),
            "fieldname": "trailer",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": _("Driver"),
            "fieldname": "assigned_driver_name",
            "fieldtype": "Data",
            "width": 215,
        },
        {
            "label": _("Fuel Warehouse"),
            "fieldname": "warehouse",
            "fieldtype": "Link",
            "options": "Warehouse",
            "width": 170,
        },
        {
            "label": _("Cost Centre"),
            "fieldname": "cost_centre",
            "fieldtype": "Link",
            "options": "Cost Center",
            "width": 150,
        },
        {
            "label": _("Load(km/L)"),
            "fieldname": "loaded_fuel_consumption",
            "fieldtype": "Float",
            "precision": 2,
            "width": 115,
        },
        {
            "label": _("Empty (km/L)"),
            "fieldname": "empty_fuel_consumption",
            "fieldtype": "Float",
            "precision": 2,
            "width": 115,
        },
        {
            "label": _("Status"),
            "fieldname": "truck_status",
            "fieldtype": "Data",
            "width": 120,
        },
    ]


def get_data(filters):
    conditions = get_conditions(filters)

    return frappe.db.sql(
        f"""
        SELECT 
            name,
            horse,
            assigned_trailer as trailer,  # Changed from trailer to assigned_trailer
            assigned_driver_name,
            warehouse,
            cost_centre,
            loaded_fuel_consumption,
            empty_fuel_consumption,
            truck_status
        FROM `tabTruck`
        WHERE 1=1
        {conditions}
        ORDER BY name
    """,
        filters,
        as_dict=1,
    )


def get_conditions(filters):
    conditions = []

    if filters.get("truck_status"):
        conditions.append("AND truck_status = %(truck_status)s")

    if filters.get("driver_linked"):
        if filters.get("driver_linked") == "Yes":
            conditions.append("AND assigned_driver IS NOT NULL")
        else:
            conditions.append("AND assigned_driver IS NULL")

    return " ".join(conditions)

File: freightmas/trucking_service/report/truck_register/truck_register.js
// Copyright (c) 2025, Zvomaita Technologies (Pvt) Ltd and contributors
// For license information, please see license.txt

frappe.query_reports["Truck Register"] = {
	"filters": [
		{
			fieldname: "truck_status",
			label: __("Status"),
			fieldtype: "Select",
			options: "\nAvailable\nNot Available"
		},
		{
			fieldname: "driver_linked",
			label: __("Driver Linked"),
			fieldtype: "Select",
			options: "\nYes\nNo"
		}
	],

	onload: function(report) {
		report.page.add_inner_button('Export to Excel', function() {
			const filters = report.get_filter_values(true);
			const query = encodeURIComponent(JSON.stringify(filters));
			const url = `/api/method/freightmas.api.export_report_to_excel?report_name=Truck Register&filters=${query}`;
			window.open(url);
		}, 'Export');

		report.page.add_inner_button('Export to PDF', function() {
			const filters = report.get_filter_values(true);
			const query = encodeURIComponent(JSON.stringify(filters));
			const url = `/api/method/freightmas.api.export_report_to_pdf?report_name=Truck Register&filters=${query}`;
			window.open(url);
		}, 'Export');

		// Add Clear Filters button
		report.page.add_inner_button('Clear Filters', function() {
			report.filters.forEach(filter => {
				let default_value = filter.df.default || "";
				report.set_filter_value(filter.df.fieldname, default_value);
			});
		});
	}
};
